helix-term/src/grammars.rs

```diff
@@ -0,0 +1,290 @@
+use anyhow::{anyhow, Context, Result};
+use std::fs;
+use std::time::SystemTime;
+use std::{
+    path::{Path, PathBuf},
+    process::Command,
+    sync::mpsc::channel,
+};
+
+use helix_core::syntax::{GrammarConfiguration, GrammarSource, DYLIB_EXTENSION};
+
+const BUILD_TARGET: &str = env!("BUILD_TARGET");
+const REMOTE_NAME: &str = "helix-origin";
+
+pub fn fetch_grammars() {
+    run_parallel(get_grammar_configs(), fetch_grammar);
+}
+
+pub fn build_grammars() {
+    run_parallel(get_grammar_configs(), build_grammar);
+}
+
+fn run_parallel<F>(grammars: Vec<GrammarConfiguration>, job: F)
+where
+    F: Fn(GrammarConfiguration) + std::marker::Send + 'static + Copy,
+{
+    let mut n_jobs = 0;
+    let pool = threadpool::Builder::new().build();
+    let (tx, rx) = channel();
+
+    for grammar in grammars {
+        let tx = tx.clone();
+        n_jobs += 1;
+
+        pool.execute(move || {
+            job(grammar);
+
+            // report progress
+            tx.send(1).unwrap();
+        });
+    }
+    pool.join();
+
+    assert_eq!(rx.try_iter().sum::<usize>(), n_jobs);
+}
+
+pub fn fetch_grammar(grammar: GrammarConfiguration) {
+    if let GrammarSource::Git { remote, revision } = grammar.source {
+        let grammar_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
+            .join("../runtime/grammars/sources")
+            .join(grammar.grammar_id.clone());
+
+        fs::create_dir_all(grammar_dir.clone()).expect("Could not create grammar directory");
+
+        // create the grammar dir contains a git directory
+        if !grammar_dir.join(".git").is_dir() {
+            Command::new("git")
+                .args(["init"])
+                .current_dir(grammar_dir.clone())
+                .output()
+                .expect("Could not execute 'git'");
+        }
+
+        // ensure the remote matches the configured remote
+        if get_repository_info(&grammar_dir, vec!["remote", "get-url", REMOTE_NAME])
+            != Some(remote.clone())
+        {
+            set_remote(&grammar_dir, &remote);
+        }
+
+        // ensure the revision matches the configured revision
+        if get_repository_info(&grammar_dir, vec!["rev-parse", "HEAD"]) != Some(revision.clone()) {
+            // Fetch the exact revision from the remote.
+            // Supported by server-side git since v2.5.0 (July 2015),
+            // enabled by default on major git hosts.
+            Command::new("git")
+                .args(["fetch", REMOTE_NAME, &revision])
+                .current_dir(grammar_dir.clone())
+                .output()
+                .expect("Failed to execute 'git'");
+
+            Command::new("git")
+                .args(["checkout", &revision])
```

Comment:
We probably want shallow and recursive submodules right? I think some of them have submodules.

Reply:
Can we also implement a fast path for github where it doesn't use git but instead downloads a zip? i.e. `https://github.com/helix-editor/helix/archive/<sha>.zip`

Reply:
Would it be alright to leave that off to a follow-up PR? Naively I think that would be a lot of new code (plus we would need an HTTP client, right?) and this PR is already pretty sprawling

Reply:
True, yeah